src/coreclr/vm/virtualcallstub.cpp

```diff
@@ -1262,6 +1474,195 @@ ResolveCacheElem* __fastcall VirtualCallStubManager::PromoteChainEntry(ResolveCa
 }
 #endif // CHAIN_LOOKUP
 
+#ifdef FEATURE_CACHED_INTERFACE_DISPATCH
+PCODE CachedInterfaceDispatchResolveWorker(StubCallSite* pCallSite, OBJECTREF *protectedObj, DispatchToken token)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+        INJECT_FAULT(COMPlusThrowOM(););
+        PRECONDITION(protectedObj != NULL);
+        PRECONDITION(*protectedObj != NULL);
+        PRECONDITION(IsProtectedByGCFrame(protectedObj));
+    } CONTRACTL_END;
+
+    MethodTable* objectType = (*protectedObj)->GetMethodTable();
+    CONSISTENCY_CHECK(CheckPointer(objectType));
+
+    PCODE target = (PCODE)NULL;
+    BOOL patch = VirtualCallStubManager::Resolver(objectType, token, protectedObj, &target, TRUE /* throwOnConflict */);
+
+#if defined(_DEBUG)
+    if (!objectType->IsComObjectType()
+        && !objectType->IsIDynamicInterfaceCastable())
+    {
+        CONSISTENCY_CHECK(!MethodTable::GetMethodDescForSlotAddress(target)->IsGenericMethodDefinition());
+    }
+#endif // _DEBUG
+
+    if (patch && pCallSite != NULL)
+    {
+        DispatchCellInfo cellInfo = ((InterfaceDispatchCell*)pCallSite->GetIndirectCell())->GetDispatchCellInfo();
+        InterfaceDispatch_UpdateDispatchCellCache((InterfaceDispatchCell*)pCallSite->GetIndirectCell(), target, objectType, &cellInfo);
+    }
+
+    return target;
+}
+
+// Resolve a dispatch on a virtual open delegate without updating any pointers
+extern "C" PCODE CID_VirtualOpenDelegateDispatchWorker(TransitionBlock * pTransitionBlock, PCODE* ppMethodPtrAuxField)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        INJECT_FAULT(COMPlusThrowOM(););
+        PRECONDITION(CheckPointer(pTransitionBlock));
+        MODE_COOPERATIVE;
+    } CONTRACTL_END;
+
+    OBJECTREF delegateObj = ObjectToOBJECTREF((Object*)(((BYTE*)ppMethodPtrAuxField) - DelegateObject::GetOffsetOfMethodPtrAux()));
+    MAKE_CURRENT_THREAD_AVAILABLE();
+
+#ifdef _DEBUG
+    Thread::ObjectRefFlush(CURRENT_THREAD);
+#endif
+
+    StubDispatchFrame frame(pTransitionBlock);
+    StubDispatchFrame * pSDFrame = &frame;
+
+    OBJECTREF *protectedObj = pSDFrame->GetThisPtr();
+    _ASSERTE(protectedObj != NULL);
+    OBJECTREF pObj = *protectedObj;
+
+    PCODE target = (PCODE)NULL;
+
+    if (pObj == NULL) {
+        pSDFrame->SetForNullReferenceException();
+        pSDFrame->Push(CURRENT_THREAD);
+        INSTALL_MANAGED_EXCEPTION_DISPATCHER;
+        INSTALL_UNWIND_AND_CONTINUE_HANDLER;
+        COMPlusThrow(kNullReferenceException);
+        UNINSTALL_UNWIND_AND_CONTINUE_HANDLER;
+        UNINSTALL_MANAGED_EXCEPTION_DISPATCHER;
+        _ASSERTE(!"Throw returned");
+    }
+
+    MethodDesc *pTargetMD = COMDelegate::GetMethodDescForOpenVirtualDelegate(delegateObj);
+    pSDFrame->SetFunction(pTargetMD);
```

Comment:
If `pSDFrame` is `&frame` why are we indirecting through the pointer instead of using `frame` directly?

Reply:
This is due to historical reasons. Until recently (3 weeks ago), the frame was `FrameWithCookie<StubDispatchFrame>` and not just `StubDispatchFrame`. So the `StubDispatchFrame * pSDFrame = &frame` was used so that on every access we don't need to do casting. It isn't needed anymore.